src/tools/ScenarioMeasurement/Startup/WinUIParser.cs

```diff
@@ -0,0 +1,87 @@
+﻿using Microsoft.Diagnostics.Tracing;
+using Reporting;
+using System;
+using System.Collections.Generic;
+
+namespace ScenarioMeasurement
+{
+    public class WinUIParser : IParser
+    {
+        public void EnableKernelProvider(ITraceSession kernel)
+        {
+            kernel.EnableKernelProvider(TraceSessionManager.KernelKeyword.Process, TraceSessionManager.KernelKeyword.Thread, TraceSessionManager.KernelKeyword.ContextSwitch);
+        }
+
+        public void EnableUserProviders(ITraceSession user)
+        {
+            user.EnableUserProvider("Microsoft-Windows-XAML", TraceEventLevel.Verbose);
+        }
+
+        public IEnumerable<Counter> Parse(string mergeTraceFile, string processName, IList<int> pids, string commandLine)
+        {
+                       var results = new List<double>();
+            var threadTimes = new List<double>();
+            double threadTime = 0;
+            var ins = new Dictionary<int, double>();
+            double start = -1;
+            int? pid = null;
+            using (var source = new ETWTraceEventSource(mergeTraceFile))
+            {
+
+                source.Kernel.ProcessStart += evt =>
+                {
+                    if (processName.Equals(evt.ProcessName, StringComparison.OrdinalIgnoreCase) && pids.Contains(evt.ProcessID) && evt.CommandLine.Trim() == commandLine.Trim())
+                    {
+                        if (pid.HasValue)
+                        {
+                            // Processes might be reentrant. For now this traces the first (outermost) process of a given name.
+                            return;
+                        }
+                        pid = evt.ProcessID;
+                        start = evt.TimeStampRelativeMSec;
+                    }
+                };
+
+                source.Kernel.ThreadCSwitch += evt =>
+                {
+                    if (!pid.HasValue) // we're currently in a measurement interval
+                        return;
+
+                    if (evt.NewProcessID != pid && evt.OldProcessID != pid)
+                        return; // but this isn't our process
+
+                    if (evt.OldProcessID == pid) // this is a switch out from our process
+                    {
+                        if (ins.TryGetValue(evt.OldThreadID, out var value)) // had we ever recorded a switch in for this thread? 
```

Comment:
According to [this guide](https://github.com/microsoft/perfview/blob/570cc0364c7e2407222f84600aebb9d9cb2f6007/documentation/TraceEvent/TraceEventProgrammersGuide.md), all these event handlers are fired asynchronously, so I think there are likely to be lots of race conditions here with how this code is set up. In addition, the `List` and `Dictionary` types are not thread-safe, and `+=` is also not atomic either.

At the bottom of the guide it says that if you are just parsing a file and are not handling live streams of events, you should use the `TraceLog` type instead as it lets you iterate over all the events in a for loop rather than needing to set up event handlers. I would say you should switch it to use that instead.

Reply:
I have learnt from another PR that my comment here was wrong and that it is thread-safe, so resolving the comment